Feature/react select single value

jvorcak opened the pull request:
This PR allows our ReactSelect to work with values instead of objects, based on the options property.

jankoritak commented on packages/ui/src/SelectField.tsx:
```diff
@@ -181,6 +182,37 @@ export const SelectField = <V,>({
+  )
+
+  const selectedOption: SelectFieldOption<V> | SelectFieldOption<V>[] | undefined = useMemo(() => {
+    if (isMulti) {
+      // if it's multi value, let's transform a value array to an array containing SelectFieldOptions
+      const valueArray = value as V[] | null
```
Suggestion: WDYT about providing a type-guard here instead of inline `as` cast?

jankoritak commented on packages/ui/src/SelectField.tsx:
```diff
@@ -181,6 +182,37 @@ export const SelectField = <V,>({
+    if (isMulti) {
+      // if it's multi value, let's transform a value array to an array containing SelectFieldOptions
+      const valueArray = value as V[] | null
+      return (
+        valueArray?.map((val) => ({
+          value: val,
+          // we need to find instead of value: object mapping so that we can support objects as values
+          label: (options.find(({ value }) => value === val) as SelectFieldOption<V>).label,
+        })) ?? []
+      )
```
Nit: We can remove these here. 

```suggestion
      return
        valueArray?.map((val) => ({
          value: val,
          // we need to find instead of value: object mapping so that we can support objects as values
          label: (options.find(({ value }) => value === val) as SelectFieldOption<V>).label,
        })) ?? []
```

jvorcak replied:
I'm working on adjusting this to a creatable option. I'll push it soon, can you please re-review afterwards?

jankoritak replied:
Sure, sry. Missed the `WIP`.

jankoritak reviewed (approved):
LGTM, leaving 2 suggestions.

aigoncharov commented on packages/ui/src/SelectField.tsx:
```diff
@@ -106,28 +106,45 @@ export type SelectFieldCoreDynamicProps<V> =
+      onChange: (newValue: V[]) => void
     }
 
+// Since the user input is string, let's allow creatable only for string
```
Good thinking!

aigoncharov commented on packages/ui/src/SelectField.tsx:
```diff
@@ -154,6 +174,43 @@ const getMenuContainer = (menuPortalTarget: 'body' | 'self' | HTMLElement | null
+      value?.map((val) => ({
+        value: val,
+        // we need to find instead of value: object mapping so that we can support objects as values
+        label: (options.find(({ value }) => value === val) as SelectFieldOption<V>)?.label ?? val,
```
nit: Should we a hashmap out of `options` for O(1) access?

aigoncharov commented on packages/ui/src/SelectField.tsx:
```diff
@@ -154,6 +174,43 @@ const getMenuContainer = (menuPortalTarget: 'body' | 'self' | HTMLElement | null
+ * @param options - an array of objects describing all possible options with their labels
+ * @param isMulti - boolean indicating whether we're in a multiple mode or not
+ */
+export function getSelectedOptionFromValue<V>({ value, options, isMulti }: GetSelectedOptionFromValueProps<V>) {
```
nit: Should it ba a const arrow function?

aigoncharov commented on packages/ui/src/SelectField.tsx:
```diff
@@ -184,6 +242,35 @@ export const SelectField = <V,>({
+    [onChange],
+  )
+
+  const onCreateOption = React.useCallback(
```
Why do we need this one? AFAIU, react-select handles this logic automatically. I tried to remove it locally and it just worked. Then we'll also be able to remove `isCreatableGuard
`. Moreover, with the current implementation, a single-value creatable select does not work.

jvorcak replied:
Good catch @aigoncharov

aigoncharov reviewed (commented):
@jvorcak great job and wonderful idea! Left a single comment to address. After that - good to go.

danad02 commented on packages/ui/src/SelectField.tsx:
```diff
@@ -154,6 +174,43 @@ const getMenuContainer = (menuPortalTarget: 'body' | 'self' | HTMLElement | null
+ * @param isMulti - boolean indicating whether we're in a multiple mode or not
+ */
+export function getSelectedOptionFromValue<V>({ value, options, isMulti }: GetSelectedOptionFromValueProps<V>) {
+  if (isMultipleModeGuard(value, isMulti)) {
+    // if it's multi value, let's transform a value array to an array containing SelectFieldOptions
+    return (
+      value?.map((val) => ({
```
I feel like that guard is not entirely correct. We tell TypeScript that value is certainly `V[]` while in fact it could also be `null`. Then we check with optional chaining if value truly is not null. If you remove the `?` from `value?.map()` TypeScript won't error while in fact it should.

danad02 commented on packages/ui/src/SelectField.tsx:
```diff
@@ -154,6 +174,43 @@ const getMenuContainer = (menuPortalTarget: 'body' | 'self' | HTMLElement | null
+export function getSelectedOptionFromValue<V>({ value, options, isMulti }: GetSelectedOptionFromValueProps<V>) {
+  if (isMultipleModeGuard(value, isMulti)) {
+    // if it's multi value, let's transform a value array to an array containing SelectFieldOptions
+    return (
+      value?.map((val) => ({
+        value: val,
+        // we need to find instead of value: object mapping so that we can support objects as values
+        label: (options.find(({ value }) => value === val) as SelectFieldOption<V>)?.label ?? val,
+      })) ?? []
```
return value.map((val) => ({
      value: val,
      // we need to find instead of value: object mapping so that we can support objects as values
      label: (options.find(({ value }) => value === val) as SelectFieldOption<V>)?.label ?? val,
    }))

danad02 commented on packages/ui/src/SelectField.tsx:
```diff
@@ -154,6 +174,43 @@ const getMenuContainer = (menuPortalTarget: 'body' | 'self' | HTMLElement | null
+export function getSelectedOptionFromValue<V>({ value, options, isMulti }: GetSelectedOptionFromValueProps<V>) {
+  if (isMultipleModeGuard(value, isMulti)) {
+    // if it's multi value, let's transform a value array to an array containing SelectFieldOptions
+    return (
+      value?.map((val) => ({
+        value: val,
+        // we need to find instead of value: object mapping so that we can support objects as values
+        label: (options.find(({ value }) => value === val) as SelectFieldOption<V>)?.label ?? val,
+      })) ?? []
+    )
```
```suggestion
    return value.map((val) => ({
      value: val,
      // we need to find instead of value: object mapping so that we can support objects as values
      label: (options.find(({ value }) => value === val) as SelectFieldOption<V>)?.label ?? val,
    }))
```

jvorcak replied:
@danad02 but value can be `null` if you use multi mode combined with creatable

danad02 replied:
@jvorcak It can't, I just tested it

danad02 replied:
@jvorcak Confirming that it's not an issue related to this PR. We have wrong types for `multi` from the beginning.

jvorcak replied:
<img width="674" alt="Screenshot 2021-03-04 at 9 40 50" src="https://user-images.githubusercontent.com/1083817/109935292-badb3480-7ccd-11eb-8f2e-b73a353895b2.png">
<img width="737" alt="Screenshot 2021-03-04 at 9 40 32" src="https://user-images.githubusercontent.com/1083817/109935309-be6ebb80-7ccd-11eb-884c-dfec1b0602a9.png">

danad02 replied:
Well yes, you got it as a default value in the `useState`. If we change typings and you use `[]` as the default value, you won't get null there.

danad02 replied:
@jvorcak Sorry to drop this non-related issue on you during this great enhancement. I just think it's a good thing to fix it sooner rather than later.

danad02 commented on packages/ui/src/SelectField.tsx:
```diff
@@ -106,28 +106,45 @@ export type SelectFieldCoreDynamicProps<V> =
+      isCreatable?: false
+    }
+
+function isMultipleModeGuard<V>(value: V | V[], isMultiple: boolean): value is V[] {
```
```suggestion
function isMultipleModeGuard<V>(value: V | V[] | null, isMultiple: boolean): value is V[] {
```

danad02 commented on packages/ui/src/SelectField.tsx:
```diff
@@ -184,6 +242,35 @@ export const SelectField = <V,>({
     }
   }, [menuPortalTarget])
 
+  const onChangeMultipleFn = React.useCallback(
+    (values: SelectFieldOption<V>[] | null) => {
+      ;(onChange as (newValue: V[] | null) => void)(values === null ? null : values.map(({ value }) => value))
+    },
+    [onChange],
+  )
```
Since empty `multi select` now returns an empty array when there's nothing selected, we might want to remove all these ugly `null`'s from other places:

```suggestion
  const onChangeMultipleFn = React.useCallback(
    (values: SelectFieldOption<V>[]) => {
      ;(onChange as (newValue: V[]) => void)(values.map(({ value }) => value))
    },
    [onChange],
  )
```

@jvorcak Can you also check other places too?

danad02 commented on packages/ui/src/SelectField.tsx:
```diff
@@ -106,28 +106,45 @@ export type SelectFieldCoreDynamicProps<V> =
       isMulti: true
-      value: SelectFieldOption<V>[] | null
-      onChange: (newValue: SelectFieldOption<V>[] | null) => void
+      value: V[] | null
+      onChange: (newValue: V[] | null) => void
```
Since empty `multi select` now returns an empty array when there's nothing selected, we might want to remove all these ugly `null`'s from other places:

```suggestion
      value: V[]
      onChange: (newValue: V[]) => void
```

danad02 reviewed (commented):
@jvorcak I left few comments.

jvorcak commented:
Thanks guys for a very great feedback. I've adjusted it, you can re-review @aigoncharov @danad02 @jankoritak 

- We now support `string` and `number` only as values
- Multiple mode can't have null, is `[]` instead (the same way as react-select`
- Direct access when transforming values
- Removed custom function for creatable mode

danad02 reviewed (approved):
Awesome! GJ :+1:

aigoncharov reviewed (approved):
LGTM!
